### PR TITLE
populate default value in unmarshal

### LINKED ISSUE #936: Close the gap of acs-engine api and ACS public document
**Is this a request for help?**:
No
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):
ISSUE

I am noticing the api model on acs-engine is not aligned with what published [here](https://docs.microsoft.com/en-us/rest/api/compute/containerservices). For example, acs-engine requires user pass in a valid MasterProfile.Count field, but the document clearly states that is not required, and be default to 1. 

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
All

**What happened**:
Once we rollout acs-engine with acs rp in the next one or two months, broader people will notice the breaking change even with the same api-version, such as 2016-09-30. The behavior will be different.

**What you expected to happen**:
User shouldn't notice we swapped the backend service. As we still conform the same api specs.

**How to reproduce it** (as minimally and precisely as possible):
Check 2016-09-30's validate function on `MasterProfile.Count`, apparently 0 is not allowed. The document allows 0, or empty.

### PR DESCRIPTION
**Pull request by rjtsdl**
**What this PR does / why we need it**:
close api definition gap to public document

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #936 

**Special notes for your reviewer**:
One sample for masterCount, if looks good, I will apply to all the other fields.

**Release note**:

```release-note
```

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/939)

### CODE REVIEW

**Comment by rjtsdl** [comment]
@JiangtianLi , i saw OrchestratorType is using `UnmarshalText`, which needs a custom type `OrchestratorType`. If you think it is ok, i want to change all of them using `UnmarshalJSON`, which does addition decoding compared with `UnmarshalText`

Then i can use OrchestratorProfile's UnmarshalJSON to do what we have now. We can eliminate a type definition.

File: pkg/api/v20170131/types.go [hunk 1]
```diff
@@ -109,6 +110,23 @@ type MasterProfile struct {
...
+// And do fields manipulation, such as populating default value
+func (m *MasterProfile) UnmarshalJSON(b []byte) error {
+	// Need to have a alias type to avoid circular unmarshal
+	type aliasMasterProfile MasterProfile
+	mm := aliasMasterProfile{}
+	if e := json.Unmarshal(b, &mm); e != nil {
+		return e
+	}
+	*m = MasterProfile(mm)
+	if m.Count == 0 {
```
Comment (rjtsdl) [inline pkg/api/v20170131/types.go]: If other fields in MasterProfile need to populate default value, it can be done inside this function.

Comment (JiangtianLi) [inline pkg/api/v20170131/types.go]: Does json.Unmarshal fill the empty field with 0?

File: pkg/api/v20170131/types.go [hunk 2]
```diff
@@ -109,6 +110,23 @@ type MasterProfile struct {
 	subnet string
 }
 
+// UnmarshalJSON unmarshal json using the default behavior
+// And do fields manipulation, such as populating default value
+func (m *MasterProfile) UnmarshalJSON(b []byte) error {
+	// Need to have a alias type to avoid circular unmarshal
+	type aliasMasterProfile MasterProfile
```
Comment (JiangtianLi) [inline pkg/api/v20170131/types.go]: why use alias?

File: pkg/api/v20170131/types.go (hunk 1 above)
Comment (rjtsdl) [inline pkg/api/v20170131/types.go]: Yes

File: pkg/api/v20170131/types.go (hunk 2 above)
Comment (rjtsdl) [inline pkg/api/v20170131/types.go]: otherwise, when you call MasterProfile.UnmarshalJSON, it causes a circular call to itself.

**Comment by weinong** [comment]
LGTM. let me know when you are ready to checkin.

**Comment by rjtsdl** [comment]
There maybe concerns on why do we need to populate default values when we have `SetPropertiesDefaults` when GenerateTemplate. 

My principle would be 
`SetPropertiesDefaults` operates on api.ContainerService, which is agnostic to api version. When the api doesn't expose all the vlab features, we use SetPropertiesDefaults to fill in vlab.ContainerService's defaults.

The populating in the unmarshal function is to fulfill the promise we made on the versioned api documents. Think about the scenario, in version 2018-09-30, we probably will switch default storageProfile to managedDisk. However, we won't change the existing versions default, which is to storageAccount. Make sense now?

**Comment by rjtsdl** [comment]
@weinong, thx for reviewing. I will do this trick for the other versions. And also, I am inclined to do this for 2017-07-01. I will have the PR ready by Monday.

**Comment by rjtsdl** [comment]
@acs-bot test this please

**Comment by rjtsdl** [comment]
@weinong , i get code ready. PTAL

**Comment by weinong** [comment]
Reviewed 2 of 2 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/acs-engine/939)*

**Comment by JiangtianLi** [comment]
@rjtsdl LGTM. Yes, we can change orchestratorType in a separate PR.

**Comment by rjtsdl** [comment]
I will just merge it. Thx

**Pull request merged**